test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,476 @@
+	tr.checkEndpointHost(c, f, tr.nodeportEndpoint, expectedZone, tr.externalClientPod.Name, checkResponses)
+}
+
+func (tr *trafficPolicyTestResources) checkEndpointInternal(c clientset.Interface, f *framework.Framework, expectedZone string, checkResponses seenPodsCheckFunc) {
+	tr.checkEndpointHost(c, f, tr.clusterIPEndpoint, expectedZone, tr.internalClientPod.Name, checkResponses)
+}
+
+// This is used with 'Cluster' traffic policies within a zone (topology hints: auto)
+func sawMoreThanOnePod(_ *trafficPolicyTestResources, seenPods map[string]string) bool {
+	return len(seenPods) > 1
+}
+
+// This is used when 'Local' traffic policy is specified to verify the expected node
+func sawOnePodOnEndpointNode(tr *trafficPolicyTestResources, seenPods map[string]string) bool {
+	if len(seenPods) != 1 {
+		return false
+	}
+	for pod := range seenPods {
+		node, ok := tr.nodesByPod[pod]
+		return ok && node == tr.endpointNode.Name
+	}
+	return false
+}
+
+// This is used with 'Cluster' traffic policies with topology hints disabled
+func sawMultipleZones(_ *trafficPolicyTestResources, seenPods map[string]string) bool {
+	zones := make(map[string]bool)
+	for _, zone := range seenPods {
+		zones[zone] = true
+	}
+	return len(seenPods) > 1 && len(zones) > 1
+}
+
+func createClientPod(podClient *e2epod.PodClient, endpoint string, namespaceName string, nodeName string, podSuffix string) *v1.Pod {
+	ginkgo.By("creating a client pod for probing the service from " + nodeName)
+	// put a pod in the single node zone to use for curling
+	podName := "curl-from-" + nodeName + podSuffix
+
+	/// turn this into a function which takes endpoint, and expected reply details
+	clientPod := e2epod.NewAgnhostPod(namespaceName, podName, nil, nil, nil, "serve-hostname")
+	nodeSelection := e2epod.NodeSelection{Name: nodeName}
+	e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+
+	ginkgo.By("Checking access to endpoint " + endpoint)
+
+	cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 0.5; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s/hostname ; echo; done`, endpoint)
+	clientPod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", cmd}
+	clientPod.Spec.Containers[0].Name = clientPod.Name
+	return podClient.CreateSync(clientPod)
+}
+
+// Create a netexec daemonset with NodePort service use client pods in each zone to verify routing
+// need at least 3 nodes, 2 in one zone and 1 in another
+// 2 in a zone to verify Local vs Cluster and sep zones to verify topology hints
+var _ = common.SIGDescribe("[Feature:Topology Hints][Feature:ServiceInternalTrafficPolicy]", func() {
+	f := framework.NewDefaultFramework("traffic-policy")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	// filled in BeforeEach
+	var c clientset.Interface
+
+	ginkgo.BeforeEach(func() {
+		c = f.ClientSet
+		e2eskipper.SkipUnlessMultizone(c)
+		e2eskipper.SkipUnlessNodeCountIsAtLeast(3)
+	})
+
+	ginkgo.Context("With hints enabled", func() {
```

Comment:
I could be missing something, These tests don't seem sufficient to actually test that topology hints are actually working, more that xTP and iTP are correctly enabled/disabled. It seems like you'd also want to find a way to check that all pods reached were within the same zone. I think with your current daemonset-based set up that would largely be reliable unless a set of nodes were unexpectedly unschedulable.

Reply:
The goal with line 391 context of tests was to verify requests are only handled by nodes within the relevant zone when hints are enabled. To verify that, checkEndpointExternal & checkEndpointInternal will verify that the responding pod is within a zone if an 'expectedZone' arg is supplied. That zone check is around line 304.

Notice the hints disabled cases do not supply an expected zone, but the 'hints enabled' cases do (where applicable).

Let me know if this still seems incorrect.

Reply:
OK, that makes sense now, it just wasn't obvious to me from reading the code. Maybe some additional comments would help? Maybe the function names could be clearer?

Reply:
As a follow up note, it's not obvious that "topology hints" will always route traffic to endpoints in the same zone. I think that's almost certainly going to be the case here where there's a daemonset with pods that have already been marked as ready _and_ we're not expecting a change to nodes. With that said, it could be helpful to clarify those assumptions here since there's a chance they may not always be true and we could end up with flakes if that's the case.